File: src/dazzle/cli/mcp.py
"""
MCP (Model Context Protocol) CLI commands.

Commands for running and managing the DAZZLE MCP server.

Note: These are top-level commands (mcp, mcp-setup, mcp-check) not a subgroup.
They are exported as individual command functions and added directly to the main app.
"""

import asyncio
from pathlib import Path

import typer

# Create a Typer app for grouping purposes, but commands will be added
# to main app directly since they're top-level commands
mcp_app = typer.Typer(
    help="MCP (Model Context Protocol) server commands.",
    no_args_is_help=True,
)


@mcp_app.command("run")
def mcp_run(
    working_dir: Path = typer.Option(  # noqa: B008
        None,
        "--working-dir",
        help="Project root directory (default: current directory)",
    ),
) -> None:
    """
    Run DAZZLE MCP server.

    Starts the MCP (Model Context Protocol) server that provides tools
    for working with DAZZLE projects from Claude Code.
    """
    from dazzle.mcp.server import run_server

    # Use provided directory or current directory
    project_root = working_dir.resolve() if working_dir else Path.cwd()

    # Run the async server
    try:
        asyncio.run(run_server(project_root))
    except KeyboardInterrupt:
        typer.echo("\nMCP server stopped.")
    except Exception as e:
        typer.echo(f"Error running MCP server: {e}", err=True)
        raise typer.Exit(code=1)


@mcp_app.command("setup")
def mcp_setup(
    force: bool = typer.Option(
        False,
        "--force",
        help="Overwrite existing MCP server config",
    ),
) -> None:
    """
    Register DAZZLE MCP server with Claude Code.

    This command registers the MCP server in your Claude Code configuration
    so that DAZZLE tools are automatically available in all projects.
    """
    from dazzle.mcp.setup import get_claude_config_path, register_mcp_server

    config_path = get_claude_config_path()
    if not config_path:
        typer.echo("Error: Could not find Claude Code config directory", err=True)
        raise typer.Exit(code=1)

    typer.echo(f"Registering MCP server at: {config_path}")

    if register_mcp_server(force=force):
        typer.echo("✅ DAZZLE MCP server registered successfully")
        typer.echo("")
        typer.echo("Next steps:")
        typer.echo("  1. Restart Claude Code")
        typer.echo("  2. Open a DAZZLE project")
        typer.echo('  3. Ask Claude: "What DAZZLE tools do you have access to?"')
    else:
        typer.echo("❌ Failed to register MCP server", err=True)
        raise typer.Exit(code=1)


@mcp_app.command("check")
def mcp_check() -> None:
    """
    Check DAZZLE MCP server status.

    Verifies that the MCP server is registered with Claude Code and
    shows available tools.
    """
    from dazzle.mcp.setup import check_mcp_server

    status = check_mcp_server()

    typer.echo("DAZZLE MCP Server Status")
    typer.echo("=" * 50)
    typer.echo(f"Status:        {status['status']}")
    typer.echo(f"Registered:    {'✓ Yes' if status['registered'] else '✗ No'}")

    if status["config_path"]:
        typer.echo(f"Config:        {status['config_path']}")

    if status["server_command"]:
        typer.echo(f"Command:       {status['server_command']}")

    if status["tools"]:
        typer.echo("")
        typer.echo(f"Available Tools ({len(status['tools'])}):")
        for tool in sorted(status["tools"]):
            typer.echo(f"  • {tool}")
    elif status["registered"]:
        typer.echo("")
        typer.echo("Tools: Unable to enumerate (MCP SDK not available)")

    if not status["registered"]:
        typer.echo("")
        typer.echo("💡 To register the MCP server, run: dazzle mcp-setup")
        raise typer.Exit(code=1)

File: cli/src/commands/mcp.ts
/**
 * mcp command - Run MCP server for Claude Code integration
 *
 * The MCP server provides context-aware tools for working with DAZZLE projects
 * from within Claude Code.
 */

import { z } from 'zod'
import type { CommandDefinition } from '../types/commands'
import { success, error, ErrorHints } from '../lib/output'
import { getPythonPath } from '../lib/python'

const McpArgs = z.object({
  'working-dir': z.string().optional().describe('Project root directory (default: current directory)'),
})

export const mcp: CommandDefinition<typeof McpArgs> = {
  name: 'mcp',
  description: 'Run MCP server for Claude Code',
  help: `
Runs the DAZZLE MCP (Model Context Protocol) server.

This server provides tools for Claude Code to interact with DAZZLE projects:
- list_modules: List DSL modules in the project
- lookup_concept: Get documentation for DSL concepts
- find_examples: Find example code patterns
- validate_dsl: Validate DSL syntax
- analyze_patterns: Analyze project patterns

The server communicates via stdio and is typically started automatically
by Claude Code based on the MCP configuration in ~/.claude/mcp_servers.json.

To register the server with Claude Code, run:
  dazzle mcp-setup
`,
  examples: [
    'dazzle mcp',
    'dazzle mcp --working-dir /path/to/project',
  ],
  args: McpArgs,

  async run(args, ctx) {
    const python = await getPythonPath()
    const cliArgs = []

    if (args['working-dir']) {
      cliArgs.push('--working-dir', args['working-dir'])
    }

    // Run MCP server - it communicates via stdio
    const proc = Bun.spawn([python, '-m', 'dazzle.mcp', ...cliArgs], {
      cwd: ctx.cwd,
      stdio: ['inherit', 'inherit', 'inherit'],
      env: {
        ...process.env,
      },
    })

    // Wait for process to exit
    const exitCode = await proc.exited

    if (exitCode !== 0) {
      return error(
        'MCP_ERROR',
        `MCP server exited with code ${exitCode}`,
        'Check the server logs for details'
      )
    }

    return success({ message: 'MCP server stopped' })
  },
}

/**
 * mcp-setup command - Register MCP server with Claude Code
 */

const McpSetupArgs = z.object({
  force: z.boolean().default(false).describe('Overwrite existing MCP server config'),
})

export const mcpSetup: CommandDefinition<typeof McpSetupArgs> = {
  name: 'mcp-setup',
  description: 'Register MCP server with Claude Code',
  help: `
Registers the DAZZLE MCP server with Claude Code.

This command updates your Claude Code configuration (usually at
~/.claude/mcp_servers.json) to include the DAZZLE MCP server.

After registration, restart Claude Code to activate the tools.
`,
  examples: [
    'dazzle mcp-setup',
    'dazzle mcp-setup --force',
  ],
  args: McpSetupArgs,

  async run(args, ctx) {
    const python = await getPythonPath()
    const cliArgs = ['mcp', 'setup']

    if (args.force) {
      cliArgs.push('--force')
    }

    // Run mcp setup via Python CLI
    const proc = Bun.spawn([python, '-m', 'dazzle', ...cliArgs], {
      cwd: ctx.cwd,
      stdio: ['inherit', 'inherit', 'inherit'],
      env: {
        ...process.env,
      },
    })

    const exitCode = await proc.exited

    if (exitCode !== 0) {
      return error(
        'MCP_SETUP_ERROR',
        `MCP setup failed with code ${exitCode}`,
        'Check the output above for details'
      )
    }

    return success({ message: 'MCP server registered' })
  },
}

/**
 * mcp-check command - Check MCP server status
 */

const McpCheckArgs = z.object({})

export const mcpCheck: CommandDefinition<typeof McpCheckArgs> = {
  name: 'mcp-check',
  description: 'Check MCP server status',
  help: `
Checks if the DAZZLE MCP server is registered with Claude Code
and shows available tools.

Use this to verify your MCP configuration is correct.
`,
  examples: [
    'dazzle mcp-check',
  ],
  args: McpCheckArgs,

  async run(args, ctx) {
    const python = await getPythonPath()

    // Run mcp check via Python CLI
    const proc = Bun.spawn([python, '-m', 'dazzle', 'mcp', 'check'], {
      cwd: ctx.cwd,
      stdio: ['inherit', 'inherit', 'inherit'],
      env: {
        ...process.env,
      },
    })

    const exitCode = await proc.exited

    if (exitCode !== 0) {
      return error(
        'MCP_CHECK_ERROR',
        `MCP check failed`,
        'Run `dazzle mcp-setup` to register the MCP server'
      )
    }

    return success({ message: 'MCP server is configured correctly' })
  },
}
